### ISSUE TITLE
Feature: Improve test coverage for scorecard-action

### ISSUE THREAD

**Issue by laurentsimon** [ossf/scorecard-action#79]
Things we should be able to test:

**=== Unit tests: P0 =====**
1. private repos
2. different trigger events: push, schedule, pull_request, workflow_dispatch, etc. Priority should be for push, schedule and pull request
3. different branches: we only support the default branch for push and schedule 
3. forks (not supported), should be added to documentation https://github.com/ossf/scorecard-action/issues/80
4. this problem https://github.com/ossf/scorecard-action/issues/106#issuecomment-1049247360
5. Capture the different threshold scores, e.g. https://github.com/ossf/scorecard-action/pull/129

The entrypoint of the action is entrypoint.sh. It expects some variables to be set, see https://github.com/ossf/scorecard-action/blob/main/entrypoint.sh#L25. These variables are the one we need to change for each unit test.

The shell script is part of a dockerfile: the dockefile defines the GitHub action, so we can build an image and call it with different parameters to create our unit tests.

The file `$GITHUB_EVENT_PATH` should contain all the data we want, except for a GitHub API bug, see https://github.com/ossf/scorecard-action/blob/main/entrypoint.sh#L38. his means we may need to create severals repos for testing purposes until the issue is resolved. 
https://github.com/ossf/scorecard-action/pull/75#pullrequestreview-867031359

Here is an example of tests done by a contributor who filed an issue https://github.com/ossf/scorecard-action/pull/75#pullrequestreview-867031359

**=== Unit tests: P1 =====**
1. https://github.com/ossf/scorecard/issues/1196 should test that command line does not change; and that the same checks are enabled with the same user's input

**=== End-to-end tests: P2 ====**
Run the workflow/upload the SARIF results to a repo and retrieve the results via an API. Verify the results shown are as expected. This is low priority, I think, because we already test SARIF generation in scorecard itself.

cc @azeemshaikh38 @naveensrinivasan anything missing?

**Comment by naveensrinivasan** [comment]
Discussed offline with @laurentsimon, Instead of writing tests for the shell script. I am proposing we move the shell script into a `golang` code, which would be easier to maintain and specifically write `tests`.

**Comment by azeemshaikh38** [comment]
@naveensrinivasan @laurentsimon fyi I have setup an e2e example test using https://github.com/ossf-tests/scorecard-check-binary-artifacts-e2e-4-binaries/pull/1. Also, update the schedule frequency to run everyday - [link](https://github.com/ossf-tests/scorecard-check-binary-artifacts-e2e-4-binaries/blob/28bd7c492bb963117b9476a506aca5872614750b/.github/workflows/scorecards.yml#L6). We need similar setups on different repo types to cover our bases such that every commit to scorecard-action gets tested.

PS: a CloudBuild trigger will automatically build and push a Docker image on every commit - gcr.io/openssf/scorecard-action.

**Comment by laurentsimon** [comment]
Thanks Azeem. How to we monitor the results of the runs?

**Comment by azeemshaikh38** [comment]
Ah, good point hadn't thought about that. Maybe we can choose to get notified about workflow run failures using [this](https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/notifications-for-workflow-runs)? If possible setup a Slack channel to get notified about such failures (we already do this for Scorecard cron job).

**Comment by laurentsimon** [comment]
anything that's feasible works. I tend to look at more emails more than at slack. Which option do you think is better? Subscribing/watching the repos? Or something else?

**Comment by azeemshaikh38** [comment]
Email/Slack either of them works for me. Slack *might* be better (even though I don't look at Slack too often) for a few reasons:

- there is a dedicated channel and it's hard for it to get lost with other noise.
- folks can easily choose to join or leave a Slack channel so only interested parties get notified.
- since we don't need a quick turnaround time on failures, even if we notice the failures after a few days, that's ok since we only care about not releasing a broken build. 

But again, either option works for me as long as we have a way to get alerted.

Watching the repos option is not scalable. We need a way to ensure that all failures get reported in a single place.

**Comment by laurentsimon** [comment]
let's try Slack then. Agreed on turn around.
I also like the idea of having other folks watch for problems if they want to :-) or even use a bot to mine the data in the future, aggregating all failures in one place.

**Comment by naveensrinivasan** [comment]
> let's try Slack then. Agreed on turn around.
> I also like the idea of having other folks watch for problems if they want to :-) or even use a bot to mine the data in the future, aggregating all failures in one place.

I agree, Slack works for everyone!

**Comment by laurentsimon** [comment]
fyi, I don't think we even need the docker image https://github.com/ossf-tests/scorecard-check-binary-artifacts-e2e-4-binaries/blob/28bd7c492bb963117b9476a506aca5872614750b/.github/workflows/scorecards.yml#L30, we could simply pin the action by `@main` directly from the action's repo.

I suppose the use of `@main` may make things harder to debug if they fail? I think it'd still be ok since the run's log contain the commit hash that was used... but probably cleaner with the docker image.

**Comment by azeemshaikh38** [comment]
Ah, I didn't realize we could do that. I understand docker better so I went with that :) Ok to move to `@main` if you are sure about its behavior.

**Comment by laurentsimon** [comment]
yes I think it should work. I use `@branch-test` when testing my PRs.

**Comment by azeemshaikh38** [comment]
SG, let's use `@main` then.

**Comment by azeemshaikh38** [comment]
I'm updating the scope of the issue here. We need these e2e tests in place before we can safely rollout the Golang scorecard-action. 

@justaugustus @rohankh532 fyi.

**Comment by justaugustus** [comment]
@laurentsimon @azeemshaikh38 @naveensrinivasan -- Would you mind updating the issue description into the following sections:

- feature parity w/ `entrypoint.sh`
- new functionality

We should descope the Golang-based rollout to expressly the things that are required for feature parity with the current entrypoint and then take new functionality as follow-up actions.

**Comment by naveensrinivasan** [comment]
Yes, I was investigating into feature parity and realized one of the things missing/removed while refactoring. 

https://github.com/ossf/scorecard-action/blob/56e33594ff29e9458da1a27f4822d1682d0968c8/entrypoint.sh#L36-L53

I will take this up.

**Comment by justaugustus** [comment]
> I will take this up.

@naveensrinivasan -- Mind if I continue working on it while you list some of the expectations in the issue description?

I've got some local work already and between that and https://github.com/ossf/scorecard-action/pull/120, I'm worried that all three of us will work on the same thing and do it in a slightly different way.

**Comment by naveensrinivasan** [comment]
> > I will take this up.
> 
> @naveensrinivasan -- Mind if I continue working on it while you list some of the expectations in the issue description?
> 
> I've got some local work already and between that and #120, I'm worried that all three of us will work on the same thing and do it in a slightly different way.

Sounds good to me. I can start investigating missing features. Would that work?

**Comment by justaugustus** [comment]
> Sounds good to me. I can start investigating missing features. Would that work?

Perfect! Appreciate it, Naveen!

**Comment by naveensrinivasan** [comment]
### Unit tests 

####  1. Use the GitHub API to address this issue 
1. - [ ] https://github.com/ossf/scorecard-action/pull/75#pullrequestreview-867031359 
1. - [ ] this problem https://github.com/ossf/scorecard-action/issues/106#issuecomment-1049247360

This is something @justaugustus is implementing. 

####  2. We only support the default branch for push and schedule

This now only checks for `pull_request`.  Do we need to check for `schedule`? 
 - [ ] Unit tests missing for this.

https://github.com/ossf/scorecard-action/blob/56e33594ff29e9458da1a27f4822d1682d0968c8/options/options.go#L154-L160

#### 3. private repos

- What do we need to test here? 

#### 4. different trigger events: push, schedule, pull_request, workflow_dispatch, etc. Priority should be for push, schedule and pull request

- What do we need to test here?

@laurentsimon  Could you please create a separate issue for each one of these and add some details so that it is easier to track?

### e2e tests

What would need to be covered in the e2e? We would need to have specifics. Having a separate issue would be helpful with specifics.

**Comment by azeemshaikh38** [comment]
So I have created a label `golang action rollout` to track this work. I have split it out into 3 issues:

- #79 (this one) - to setup unit tests and e2e test infrastructure.
- #132 - to achieve feature parity of Golang Action with current bash action.
- #133 - to track work on Scorecard badges

Hope that helps.

**Comment by azeemshaikh38** [comment]
@naveensrinivasan for your question on e2e tests:

We need to setup repos (or use existing ones) in [ossf-tests](https://github.com/ossf-tests) org for varying scenarios (different trigger conditions, private repos, non-main default branch name etc.) and implement https://github.com/ossf/scorecard-action/issues/79#issuecomment-1049326064 so that failures result in notification on a Slack channel.

**Comment by laurentsimon** [comment]
@naveensrinivasan does @azeemshaikh38 reply answer the question?

**Comment by naveensrinivasan** [comment]
> @naveensrinivasan does @azeemshaikh38 reply answer the question?

Yes.

**Comment by naveensrinivasan** [comment]
https://github.com/ossf/scorecard-action/pull/142

**Comment by naveensrinivasan** [comment]
Closing it as it has been addressed.
